drivers/sensor/st/lis3mdl/lis3mdl.c

```diff
@@ -172,17 +174,73 @@ int lis3mdl_init(const struct device *dev)
 	return 0;
 }
 
-#define LIS3MDL_DEFINE(inst)									\
-	static struct lis3mdl_data lis3mdl_data_##inst;						\
-												\
-	static struct lis3mdl_config lis3mdl_config_##inst = {					\
-		.i2c = I2C_DT_SPEC_INST_GET(inst),						\
-		IF_ENABLED(CONFIG_LIS3MDL_TRIGGER,						\
-			   (.irq_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, irq_gpios, { 0 }),))	\
-	};											\
-												\
-	SENSOR_DEVICE_DT_INST_DEFINE(inst, lis3mdl_init, NULL,					\
-			      &lis3mdl_data_##inst, &lis3mdl_config_##inst, POST_KERNEL,	\
-			      CONFIG_SENSOR_INIT_PRIORITY, &lis3mdl_driver_api);		\
+
+
+/*
+ * Device creation macro, shared by LIS3MDL_DEFINE_SPI() and
+ * LIS3MDL_DEFINE_I2C().
+ */
+
+#define LIS3MDL_DEVICE_INIT(inst)					\
+		\
+	SENSOR_DEVICE_DT_INST_DEFINE(inst,				\
+			    lis3mdl_init,				\
+			    NULL,		\
+			    &lis3mdl_data_##inst,			\
+			    &lis3mdl_config_##inst,			\
+			    POST_KERNEL,				\
+			    CONFIG_SENSOR_INIT_PRIORITY,		\
+			    &lis3mdl_driver_api);
+
+#ifdef CONFIG_LIS3MDL_TRIGGER
+#define LIS3MDL_CFG_IRQ(inst) \
+		.irq_gpio = GPIO_DT_SPEC_INST_GET(inst, irq_gpios, { 0 }),
+#else
+#define LIS3MDL_CFG_IRQ(inst)
+#endif /* CONFIG_LIS3MDL_TRIGGER */
+
+/*
+ * Instantiation macros used when a device is on a SPI bus.
+ */
+
+#define LIS3MDL_CONFIG_SPI(inst)						\
+	{													\
+		.bus_init = lis3mdl_spi_init,					\
+		.bus_cfg.spi = SPI_DT_SPEC_INST_GET(inst, SPI_WORD_SET(8) |	\
+				      SPI_OP_MODE_MASTER |				\
+				      SPI_MODE_CPOL |					\
+				      SPI_MODE_CPHA, 0),				\
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, irq_gpios),		\
+		(LIS3MDL_CFG_IRQ(inst)), ())					\
+	}
+
+#define LIS3MDL_DEFINE_SPI(inst)							\
+	static struct lis3mdl_data lis3mdl_data_##inst;			\
+	static struct lis3mdl_config lis3mdl_config_##inst =	\
+		LIS3MDL_CONFIG_SPI(inst);							\
+	LIS3MDL_DEVICE_INIT(inst)
+
+/*
+ * Instantiation macros used when a device is on an I2C bus.
+ */
+
+#define LIS3MDL_CONFIG_I2C(inst)					\
+	{								\
+		.bus_init = lis3mdl_i2c_init,				\
+		.bus_cfg.i2c = I2C_DT_SPEC_INST_GET(inst),		\
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, irq_gpios),	\
+		(LIS3MDL_CFG_IRQ(inst)), ())				\
+	}
+
+#define LIS3MDL_DEFINE_I2C(inst)					\
+	static struct lis3mdl_data lis3mdl_data_##inst;			\
+	static struct lis3mdl_config lis3mdl_config_##inst =	\
+		LIS3MDL_CONFIG_I2C(inst);				\
+	LIS3MDL_DEVICE_INIT(inst)
```

Comment:
this macro is common between i2c and spi, so maybe you can call only once in ```LIS3MDL_DEFINE()``` as some other st driver do.